pennylane/tracker.py

```diff
@@ -0,0 +1,166 @@
+        With backpropagation, this function should take ``qnode.device``
+        instead of the device used to create the QNode.
+
+    Users can pass a custom callback function to the ``callback`` keyword. This
+    function is run each time the ``record()`` method is called. The
+    function passed must accept ``totals``, ``history``, and ``latest`` as
+    keyword arguments:
+
+    >>> def shots_info(totals=dict(), history=dict(), latest=dict()):
+    ...     print("Total shots: ", totals['shots'])
+    >>> with qml.Tracker(circuit.device, callback=shots_info) as tracker:
+    ...     qml.grad(circuit)(0.1)
+    Total shots:  100
+    Total shots:  200
+    Total shots:  300
+
+    By specifying ``persistent=False``, you can reuse the same tracker accross
+    multiple contexts.
+
+    >>> with qml.Tracker(circuit.device, persistent=False) as tracker:
+    ...     circuit(0.1)
+    >>> with tracker:
+    ...     circuit(0.2)
+    >>> tracker.totals['executions']
+    2
+
+    """
+
+    def __init__(self, dev=None, callback=None, persistent=False):
+        self.persistent = persistent
+
+        self.callback = callback
+
+        # same code as self.reset
+        self.totals = dict()
+        self.history = dict()
+        self.latest = dict()
+
+        self.tracking = False
+
+        if dev is not None:
+            if not dev.capabilities().get("supports_tracker", False):
+                raise Exception(f"Device '{dev.short_name}' does not support device tracking")
+            dev.tracker = self
+
+    def __enter__(self):
+        if not self.persistent:
+            self.reset()
+
+        self.tracking = True
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.tracking = False
+
+    def update(self, **kwargs):
+        """Store passed keyword-value pairs into ``totals``,``history``, and ``latest`` attributes.
+
+        There is no restriction on the key-value pairs passed.
+
+        >>> tracker.update(a=1, b=2)
+        >>> tracker.latest
+        {"a":1, "b":2}
+        >>> tracker.history
+        {"a": [1], "b": [2]}
+        >>> tracker.totals
+        {"a": 1, "b": 2}
+
+        """
+        self.latest = kwargs
+
+        for key, value in kwargs.items():
+            # update history
+            if key in self.history.keys():
+                self.history[key].append(value)
+            else:
+                self.history[key] = [value]
+
+            # updating totals
+            if value is not None:
+                # Only total numeric values
+                try:
+                    self.totals[key] = value + self.totals.get(key, 0)
+                except TypeError:
+                    pass
+
+    def reset(self):
+        """Resets stored information."""
+        self.totals = dict()
+        self.history = dict()
+        self.latest = dict()
+
+    def record(self):
+        """If a ``callback`` is passed to the class upon initialization, it is called."""
```

Comment:
Not sure if:
* the name of the method
* the docstring

describe this part of the code well and in its entirety.